File: 271_encode_decode_strings/encode_decode_strings.py
from typing import List

class Solution:
    def encode(self, strs: List[str]) -> str:
        res = ""
        for s in strs:
            res += str(len(s)) + "#" + s
        
        return res

    def decode(self, s: str) -> List[str]:
        res = []
        i = 0
        n = len(s)
        while i < n:
            j = i
            while j < n and s[j] != '#':
                j += 1
            length = int(s[i:j])
            j += 1
            word = s[j:j+length]
            res.append(word)
            i = j + length
        return res


if __name__ == "__main__":
    codec = Solution()
    assert codec.decode(codec.encode([])) == []
    assert codec.decode(codec.encode([""])) == [""]
    assert codec.decode(codec.encode(["", ""])) == ["", ""]
    assert codec.decode(codec.encode(["lint","code","love","you"])) == ["lint","code","love","you"]
    assert codec.decode(codec.encode(["a#b", "#", "###"])) == ["a#b", "#", "###"]
    assert codec.decode(codec.encode(["hola", "  espacios  ", "línea\nnueva"])) == ["hola", "  espacios  ", "línea\nnueva"]

File: 271_encode_decode_strings/encode_decode_strings.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type Solution struct {
}

func (c *Solution) Encode(strs []string) string {
	var builder strings.Builder

	for _, s := range strs {
		builder.WriteString(strconv.Itoa(len(s)))
		builder.WriteByte('#')
		builder.WriteString(s)
	}

	return builder.String()
}

func (c *Solution) Decode(s string) []string {
	var res []string
	n := len(s)
	i := 0

	for i < n {
		j := i
		for j < n && s[j] != '#' {
			j++
		}

		length, _ := strconv.Atoi(s[i:j])

		j++

		word := s[j : j+length]
		res = append(res, word)

		i = j + length
	}

	return res
}

func main() {
	solution := Solution{}

	tests := [][]string{
		{},
		{""},
		{"", ""},
		{"lint", "code", "love", "you"},
		{"a#b", "#", "###"},
		{"hola", "  espacios  ", "línea\nnueva"},
	}

	for _, test := range tests {
		encoded := solution.Encode(test)
		decoded := solution.Decode(encoded)
		fmt.Printf("Original: %q\nEncoded: %q\nDecoded: %q\n\n", test, encoded, decoded)
	}
}
